Unicam repacking

6by9 opened the pull request:
Adds support to the Unicam driver to allow repacking the data out of the weird CSI2 formats into the bottom of 16bit words where the relevant V4L2 formats are defined.

Also drops the table of "active_formats" enumerated at probe for enumerating the formats from the sensor driver sensor every time. This solves the problem of Bayer sensors often changing Bayer order with H & V flips.

6by9 committed:
media: bcm2835-unicam: Add support for raw14 formats

The V4L2 has gained defines for V4L2_PIX_FMT_Sxxxx14P,
therefore add support for it to bcm2835-unicam.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
media: bcm2835-unicam: Rework to not cache the list of active fmts

Some sensors will change Bayer order based on H & V flips,
therefore collecting the list of formats at async_bound has
problems.

Enumerate the formats from the sensor every time.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
media: bcm2835-unicam: Support unpacking CSI format to 16bpp

The CSI packed formats are not the easiest to work with, and
the peripheral supports unpacking them to 16bpp (but NOT
shifting the data up into the MSBs).
Where V4L2 exposes a pixfmt for both packed and unpacked
formats advertise both as being supported, and unpack the
data in the peripheral.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
media: bcm2835-unicam: Add support for luma/greyscale formats

Add support for V4L2_PIX_FMT_GREY, V4L2_PIX_FMT_Y10P,
V4L2_PIX_FMT_Y10, and V4L2_PIX_FMT_Y12 image formats from the
appropriate mediabus formats advertised by sensors.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
RFC tag added awaiting confirmation that this works with the Infineon IRS1125 driver in #3261 .

Coimbra1984 commented:
> RFC tag added awaiting confirmation that this works with the Infineon IRS1125 driver in #3261 .

Sorry for the delay (but fifa20 is out :)
This PR works perfectly for me.

Just to double check. I only built module bcm2835-unicam of your forked kernel (branch unicam-repacking) and replaced the module on my 4.19.75er kernel on a raspberry pi 3.

6by9 commented:
Thanks for the confirmation it works. RFC tag removed.

Yes, the patch only updates bcm835-unicam. As long as that was rebuilt correctly then changing the one module and rebooting should be sufficient.

Coimbra1984 commented:
> Thanks for the confirmation it works. RFC tag removed.
> 
> Yes, the patch only updates bcm835-unicam. As long as that was rebuilt correctly then changing the one module and rebooting should be sufficient.

Yes it was definitely rebuilt and I also rebooted.

6by9 commented:
@pelwell Any chance of a review?

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -455,50 +453,33 @@ static int find_mbus_depth_by_code(u32 code)
 static inline unsigned int bytes_per_line(u32 width,
-					  const struct unicam_fmt *fmt)
+					  const struct unicam_fmt *fmt,
+					  u32 v4l2_fourcc)
```
Is the extra parameter future-proofing or an abandoned idea? It doesn't seem to belong here.

6by9 replied:
In the wrong patch. Should have been in "media: bcm2835-unicam: Support unpacking CSI format to 16bpp" where the clause becomes 
```
if (v4l2_fourcc == fmt->repacked_fourcc)
		/* Repacking always goes to 16bpp */
		return ALIGN(width << 1, BPL_ALIGNMENT);
	else
		return ALIGN((width * fmt->depth) >> 3, BPL_ALIGNMENT);
```
Will fix.

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -726,14 +708,37 @@ static int unicam_enum_fmt_vid_cap(struct file *file, void  *priv,
-		return -EINVAL;
 
-	fmt = &dev->active_fmts[f->index];
+	for (i = 0; !ret && i < 128 ; i++) {
```
What's the significance of 128?

6by9 replied:
Just a failsafe to ensure a dodgy sensor driver doesn't leave us looping forever. I'll add a comment.

pelwell commented on drivers/media/platform/bcm2835/bcm2835-unicam.c:
```diff
@@ -726,14 +708,37 @@ static int unicam_enum_fmt_vid_cap(struct file *file, void  *priv,
 	const struct unicam_fmt *fmt = NULL;
+	int index = 0;
+	int ret = 0;
+	int i;
```
Nuke the following blank line.

6by9 replied:
Hmm, how did checkpatch not complain on that one?!

6by9 commented:
Branch rebased, fixes applied, an force pushed

6by9 commented:
There are no actual code changes, just moved where v4l2_fourcc was added to bytes_per_line to the right patch, and added a comment where the extra line was.
